### PR TITLE
Add library setup and teardown APIs

### PR DESCRIPTION
**Pull request by andresag01**
Add the following two functions to allow platform setup and teardown
operations for the full library to be hooked in:

* mbedtls_platform_setup()
* mbedtls_platform_teardown()

A user-configurable mbedtls_platform_context C structure is added.

Finally, the macro MBEDTLS_PLATFORM_SETUP_TEARDOWN_ALT is also added to allow
mbedtls_platform_context, mbedtls_platform_setup() and mbedtls_platform_teardown() to be overwritten by the user as required.

### CODE REVIEW

File: include/mbedtls/platform.h [hunk 1]
```diff
@@ -288,6 +288,45 @@ int mbedtls_platform_set_nv_seed(
...
+#endif /* !MBEDTLS_PLATFORM_SETUP_TEARDOWN_ALT */
+
+/**
+ * \brief   Perform any platform initialisation operations
+ *
+ * \param   ctx     mbed TLS context
+ *
+ * \return  0 if successful
+ *
+ * \note    This function should be called before any other library function
```
Comment (sbutcher-arm) [inline include/mbedtls/platform.h]: The comment needs revising. I would propose you use something similar to the following text:

```
 * \note    This function is intended to allow platform specific initialisation,
            and should be called before any other library functions. Its
            implementation is platform specific, and by default, unless platform
            specific code is provided, it does nothing.

            It's use and whether it's necessary to be called is dependent on the
            platform.
```

File: include/mbedtls/platform.h [hunk 2]
```diff
@@ -288,6 +288,45 @@ int mbedtls_platform_set_nv_seed(
...
+/**
+ * \brief   Perform any platform teardown operations
+ *
+ * \param   ctx     mbed TLS context
+ *
+ * \return  0 if successful
+ *
+ * \note    This function should be after every other mbed TLS module has been
+ *          correctly freed using the appropriate free function.
+ */
```
Comment (sbutcher-arm) [inline include/mbedtls/platform.h]: I think you should expand the comment to say the function is platform specific, similar to the comment above.

**Comment by dvir2000** [comment]
The platform initialization code seems clean and straight forward.
The only thing I would expect to be added are the calls to the setup and tear-down functions in the tests and programs code.
This would be the only way to properly test the HW acceleration integration into mbedTLS.

**Comment by RonEld** [comment]
One concern I have, is that the context given to the platform initialization\termination is platform specific, and the platform should be able to get the context in some `getter` function, or something of the sort, for it to use it whenever needed.
For example, one platform would hold handles for the HW clocks to be used for crypto accelerations, another platform would hold some initial context as part of the platform context.
AN idea of mine, which involves changing APIs :( is having the cryptographic `init` functions receive the platform context as well, and thus the alternative implementation can bind the platform context to the crypto context, if needed

I don't think this should block the current PR, but I think that it should be mentioned. This PR is pending tests to add the platform setup\teardown APIs

**Comment by sbutcher-arm** [comment]
@RonEld - Can you please raise the idea of the `getter` function as a GitHub issue, to ensure we follow up on it? It sounds like a good idea, and pretty valid concern to me.

**Comment by sbutcher-arm** [comment]
Reviewed, approved the PR, although I took the liberty of adding in my suggestions about comments in the headers, and merged.

**Pull request closed without merging**